Update state file guidance to mention cloud and remote backends, and why you shouldn't store state in VCS (#2226)
hashicorp/web-unified-docs: content/terraform/v1.15.x (rc)/docs/language/state/index.mdx

# State

Terraform must store state about your workspace's managed infrastructure and
configuration. Terraform uses your workspace's state to map real world resources
to your configuration, keep track of metadata, and to improve performance for
large infrastructures.

Terraform uses state to determine which changes to make to your infrastructure.
Prior to any operation, Terraform does a
[refresh](/terraform/cli/commands/refresh) to update the state with the real
infrastructure.

The primary purpose of Terraform state is to store bindings between objects in a
remote system and resource instances declared in your configuration. When
Terraform creates a remote object in response to a change of configuration, it
will record the identity of that remote object against a particular resource

> instance, and potentially update or delete that object in response to future

Reviewer: ```suggestion
records the identity of that remote object against a particular resource
instance, and potentially updates or deletes that object in response to future
```

configuration changes.

For more information on why Terraform requires state and why Terraform cannot
function without state, please see the page [state purpose](/terraform/language/state/purpose).

## Storing State

By default, Terraform stores each workspace's state in a local file named
`terraform.tfstate`, and a backup of the previous state in
`terraform.tfstate.backup`. Storing state locally requires no additional
configuration, but limits the ability to collaborate with others, and risks
losing workspace state if the local state file is lost. We recommend storing
state in [HCP Terraform](/terraform/cloud-docs/migrate) or a [remote
backend](/terraform/language/backend) to securely store state and collaborate
with team members.

> Do not store your state in a Version Control System or other storage solution

Reviewer: ```suggestion
Never store your state in a version control system or other storage solution
```
How strong do we need to be in this guidance? Should we say "never"?

> that does not support Terraform state locking and secure access control, because

Reviewer 2: > Do not store your state in a Version Control System or other storage solution storage solutions that does not support Terraform state locking and secure access control...

Kind of a nit, but there's several backends that we officially support that do not support native state locking. Maybe we say something "use a backend that uses state locking if possible" or something along those lines?
